src/hotspot/share/opto/parseHelper.cpp

```diff
@@ -617,67 +667,51 @@ Node* PEAState::materialize(GraphKit* kit, Node* var) {
 #endif
 
   if (oop_type->isa_instptr()) {
-    ciInstanceKlass* ik = oop_type->is_instptr()->instance_klass();
+    // virt->_oop_type is an exact non-null pointer. oop_type may not be exact, or BOT
+    // We check that they both refer to the same java type.
+    assert(virt->_oop_type->is_instptr()->is_same_java_type_as(oop_type), "type of oopptr is inconsistent!");
 #ifndef PRODUCT
     if (PEAVerbose) {
+      ciInstanceKlass* ik = oop_type->is_instptr()->instance_klass();
       tty->print("ciInstanceKlass: ");
       ik->print_name_on(tty);
       tty->cr();
     }
 #endif
 
-    for (int i = 0; i < ik->nof_nonstatic_fields(); ++i) {
-      ciField* field = ik->nonstatic_field_at(i);
-      BasicType bt = field->layout_type();
-      const Type* type = Type::get_const_basic_type(bt);
-      bool is_obj = is_reference_type(bt);
-      Node* val = virt->get_field(i);
+    for (auto&& it = virt->field_begin(); it != virt->field_end(); ++it) {
+      ciField* field = it.field();
+      Node* val = it.value();
 
 #ifndef PRODUCT
       if (PEAVerbose) {
-        tty->print("flt#%2d: ", i);
+        tty->print("field: ");
         field->print_name_on(tty);
         tty->cr();
       }
 #endif
       // no initial value or is captured by InitializeNode
       if (val == nullptr) continue;
 
-      if (is_obj && pea->is_alias(val)) {
-        // recurse if val is a virtual object.
-        if (as_virtual(pea, val)) {
-          materialize(kit, val);
-        }
-        EscapedState* es = as_escaped(pea, val);
-        assert(es != nullptr, "the object of val is not Escaped");
-        val = es->merged_value();
-      }
-
-      int offset = field->offset_in_bytes();
-      Node* adr = kit->basic_plus_adr(objx, objx, offset);
-      const TypePtr* adr_type = C->alias_type(field)->adr_type();
-      DecoratorSet decorators = IN_HEAP;
+      put_field(kit, field, objx, val);
+    }
 
-      // Store the value.
-      const Type* field_type;
-      if (!field->type()->is_loaded()) {
-        field_type = TypeInstPtr::BOTTOM;
-      } else {
-        if (is_obj) {
-          field_type = TypeOopPtr::make_from_klass(field->type()->as_klass());
-        } else {
-          field_type = Type::BOTTOM;
+    // back from DFS, we still need to check again for all virtual states.
```

Comment:
I'm ok with this for now so we can pass regression tests. In the future, I think we will need to revisit our PEA design. Iterating over every object every time we materialize can maybe be expensive.

Reply:
yes, we may have better data structure for this, but I think current move is acceptable.   We only iterate all objects of the current allocation state. It only tracks live objects of current basic block. 

Actually, I found that Graal has a similar look backward for each materialization. 
https://github.com/oracle/graal/blob/87e0f80b1b3c1621e1e0c9f4a3b0f27816b0142f/compiler/src/jdk.internal.vm.compiler/src/org/graalvm/compiler/virtual/phases/ea/PartialEscapeClosure.java#L554